src/prof_recent.c

```diff
@@ -506,33 +519,106 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
 	emitter_init(&emitter, emitter_output_json_compact,
 	    buf_writer_get_write_cb(&buf_writer),
 	    buf_writer_get_cbopaque(&buf_writer));
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
 	emitter_begin(&emitter);
 
 	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 	prof_recent_alloc_assert_count(tsd);
+	prof_recent_t *dump_dummy = prof_recent_alloc_dummy;
+	assert(dump_dummy != NULL);
+	prof_recent_alloc_dummy = prof_recent_alloc_dummy_reserve;
+	assert(prof_recent_alloc_dummy != NULL);
+	assert(prof_recent_alloc_dummy->next == prof_recent_alloc_dummy);
+	const ssize_t dump_count = prof_recent_alloc_count;
+	prof_recent_alloc_count = 0;
+	const ssize_t dump_max = prof_recent_alloc_max_get(tsd);
+	prof_recent_alloc_assert_count(tsd);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 
-	/*
-	 * Set prof_recent_alloc_max to 0 so that dumping won't block sampled
-	 * allocations: the allocations can complete but will not be recorded.
-	 */
-	ssize_t max = prof_recent_alloc_max_update(tsd, 0);
+	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize,
+		&dump_max);
+	emitter_json_array_kv_begin(&emitter, "recent_alloc");
 
-	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &max);
+	buf_writer_set_manual(&buf_writer);
+	prof_recent_t *node = NULL;
+	prof_recent_t *last = NULL;
+	while (true) {
+		malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		if (node == NULL) {
+			node = dump_dummy->next;
+		}
+		assert(node != NULL);
+		bool need_more_batch = false;
+		while (node != dump_dummy) {
+			buf_writer_commit(&buf_writer);
+			prof_recent_alloc_dump_node(&emitter, node);
+			if (buf_writer_overflow(&buf_writer)) {
+				buf_writer_revert(&buf_writer);
+				if (buf_writer_empty(&buf_writer)) {
+					buf_writer_set_auto(&buf_writer);
+					prof_recent_alloc_dump_node(
+					    &emitter, node);
+					buf_writer_set_manual(&buf_writer);
```

Comment:
Actually, when I'm looking at it again, I agree that the auto/manual naming doesn't sound that intuitive. Perhaps let's get the naming right first? How about `set_auto_flush` and `set_manual_flush`?

Reply:
Is it true that manual flush doesn’t allow silent flushing, and needs to trigger flush explicitly? If so, maybe one function to take a bool input which is whether flushing is allowed? _allow_flush?

Reply:
But even after calling `buf_writer_set_allow_flush(false)`, as you said, we can still explicitly call `buf_writer_flush()`, which seems to invalidate the `buf_writer_set_allow_flush` naming.

How about `buf_writer_set_allow_auto_flush()`, or `buf_writer_set_allow_implicit_flush()`? I think `auto` is better than `implicit`, because `implicit_flush` makes me feel that what I write is all implicitly flushed, which is not true.

Reply:
I see your point. To make it tiny bit shorter, `buf_writer_allow_auto_flush()`?

Reply:
`buf_writer_allow_auto_flush()` sounds like returning a boolean of whether the buffered writer currently allows auto flush or not, which can actually be a useful utility function to have, not for this usage case but perhaps for some other cases in the future. How about still including the `set`?

Reply:
SGTM.